content/src/service/snapshots/SnapshotManager.ts

```diff
@@ -103,6 +184,135 @@ export class SnapshotManager {
+      entity_type: entityType.toString()
+    })
+
+    try {
+      const previousHash = this.lastSnapshotsPerEntityType.get(entityType)?.hash
+
+      // Write to tmpFile the snapshot obtained
+      const { snapshotTimestamp, elements, timeElapsed } = await this.writeToFile(tmpFile, entityType)
+
+      // Hash the snapshot
+      const hash = await hashStreamV1(fs.createReadStream(tmpFile) as any)
+
+      // if success move the file to the contents folder
+      await this.moveSnapshotFileToContentFolder(tmpFile, { hash, snapshotTimestamp, elements, timeElapsed })
+
+      // Save the snapshot hash and metadata
+      const newSnapshot = { hash, lastIncludedDeploymentTimestamp: snapshotTimestamp }
+      this.lastSnapshotsPerEntityType.set(entityType, newSnapshot)
+
+      // Delete the previous full snapshot (if it exists)
+      this.removePreviousSnapshotFile(previousHash)
+    } catch (err: any) {
+      stopTimer({ failed: 'true' })
+      SnapshotManager.LOGGER.error(err)
+    } finally {
+      stopTimer({ failed: 'false' })
+      await this.deleteStagingFile(tmpFile)
+    }
+  }
+
+  private async deleteStagingFile(tmpFile: string) {
+    if (await checkFileExists(tmpFile)) {
+      try {
+        await fs.promises.unlink(tmpFile)
+      } catch (err) {
+        SnapshotManager.LOGGER.error(err)
+      }
+    }
+  }
+
+  private removePreviousSnapshotFile(previousHash: string | undefined) {
+    // the deletion of the files is deferred two minutes because there may be peers
+    // still using the content files
+    setTimeout(() => {
+      if (previousHash && this.shouldPrunePreviousSnapshot(previousHash)) {
+        this.service.deleteContent([previousHash]).catch(SnapshotManager.LOGGER.error)
+      }
+    }, 2 * 60000)
+  }
+
+  private async moveSnapshotFileToContentFolder(
+    tmpFile: string,
+    options: {
+      hash: string
+      snapshotTimestamp: number
+      elements: number
+      timeElapsed: number
+    }
+  ) {
+    const destinationFilename = path.resolve(this.components.staticConfigs.contentStorageFolder, options.hash)
+
+    const hasContent = await this.service.getContent(options.hash)
+
+    if (!hasContent) {
+      // move and compress the file into the destinationFilename
+      await this.service.storeContent(options.hash, fs.createReadStream(tmpFile))
+      SnapshotManager.LOGGER.info(
+        `Generated snapshot. hash=${options.hash} lastIncludedDeploymentTimestamp=${options.snapshotTimestamp} elements=${options.elements} timeElapsed=${options.timeElapsed}`
+      )
+      await compressContentFile(destinationFilename)
+    } else {
+      SnapshotManager.LOGGER.debug(
+        `Snapshot didn't change. hash=${options.hash} lastIncludedDeploymentTimestamp=${options.snapshotTimestamp} elements=${options.elements} timeElapsed=${options.timeElapsed}`
+      )
+    }
+  }
+
+  private async writeToFile(tmpFile: string, entityType: EntityType | ALL_ENTITIES) {
```

Comment:
I see `writeToFile` a bit confusing and it seems to be an important method, what do you think about renaming it to something like `writeDeploymentsToFile`?